### PR TITLE
Made working with 32bit KSP.exe + cleanup

### PR DESCRIPTION
**Pull request by ghost**
- removed this-prefixes, used shortcuts like get => value, properties start with big letter (PascalCasing)
- main logic/change in Start() - will try 64bit first, then current 32bit with type check, full search with printout as last option
- fix for NullReference reworked: DefaultVessels can return null, Start will set delaySetMode = true for next Update()
- debug build includes more logs (see Log() with Conditional("DEBUG") at the end)

note: I am not completely sure when the list is trully initialized, I left commented-out code there (in DefaultVessels) that can build the list of vessels itself, but it seems to me that it is better to just wait and do the job in next Update()

P.S.: Consider allowing Update() to use delaySetMode in some later call when DefaultVessels != null
(the main reason for the NullReference was that Start() attempted to set the mode too early, when DefaultVessels == null, it crashed then before setting render = true - I assume that ManeuverQueue.Start() can sometimes be called before SpaceTracking.Start())

### CODE REVIEW

File: ManeuverQueue_KSP1.2/ManeuverQueue.cs [hunk 1]
```diff
@@ -83,143 +93,239 @@ protected List<Vessel> defaultVessels
...
+			GameEvents.OnMapViewFiltersModified.Add(onMapViewFiltersModified);
 
-			this.currentMode = (FilterMode)this.pluginConfiguration.GetValue(ManeuverQueue.configurationModeKey, (int)FilterMode.Default);
+			if (DefaultVessels != null)
+			{
+				delaySetMode = false;
+				CurrentMode = (FilterMode)pluginConfiguration.GetValue(configurationModeKey, (int)FilterMode.Default);
+			}
+			else
+			{
```
Comment (ghost) [inline ManeuverQueue_KSP1.2/ManeuverQueue.cs:301]: Possibly always use delaySetMode (or StartCoroutine).
Can Unity call SpaceTracking.Start() and ManeuverQueue.Start() in parallel? That would be total mess.

**Comment by linuxgurugamer** [comment]
The code looks good, thanks.
Regarding your comment about SpaceTracking and ManeuverQueue Start, it's not Unity, but rather KSP.  it is conceviable that they can start in parallel, but I've never seen any issue like that before, so for now it can be ignored.

I am making one change, I moved the code:
```
delaySetMode = false;
CurrentMode = (FilterMode)pluginConfiguration.GetValue(configurationModeKey, (int)FilterMode.Default);
```
into a new function and calling that rather than having it duplicated.

You seem to have a good grasp on this, go ahead and move it into a CoRoutine if you like, with a yield return of 0.1 sec.

I'm going to have to replace the reflection code with something similar to EEX for several reasons, among them was a very strong request from Squad about using offsets rather than names, specifically because of the platform issue.  Keep in mind that while you are looking for  the \x1 and \x3, those can change in the future.  But for now, it looks ok.

In general, I'll have to write a method which dumps all the values for SpaceTracking and VesselIconSprites to the log when in DEBUG mode, then look at the log to get the values.

**Comment by linuxgurugamer** [comment]
I've pushed my current changes here, so if you do want to make any other changes, be sure to pull them down.

**Comment by ghost** [comment]
I will look closely at the code (both this and EEX). By offset you mean index in GetFields()? That reminds me, that .NET2.0+ uses reflection cache and you can get into troubles unless you sort the array by MetadataToken first ;) ....try fetching some field (but not the first one) then all the fileds and repeat, you won't get the same order next time!

I have described my approach already:
1. Try full names
2. Try obfuscated names ("\x1" etc.) and check the type
3. Search them all by type and print them all to log (this won't be useful for all types, but those two happen to have single field of the type that was needed)

Do you have link to where Squad gives such advices like the one about offsets? Thx.

**Comment by linuxgurugamer** [comment]
I just pushed the code for the offsets, this will now work in both 32 & 64 bit, Windows, OSX and Linux
You would have to read the thread for EEX where it was strongly suggested that I use offsets because of this issue, somewhere 1 1/2 to 2 years ago

**Comment by linuxgurugamer** [comment]
The problem with the \x1 is that there is no guarantee that the obfuscated names won't change in the future.  Using offsets eliminates that problem, with the cost of needing to update it for each new KSP version.  A small price to pay for guaranteed compatibility.

**Comment by ghost** [comment]
The problem with offsets is exactly the same as with obfuscated names, no guarantee, needs to be updated for each new KSP version. I repeat: what I did on top of it is type checking, that has chance to survive much much longer (but is quite specific for this plugin, becase there are currently no more fields of same type out of those needed).

It is your code again, do as you see fit.

BTW: Did you try the test I suggested? You really should sort the fields by MetadataToken or maybe even match the token instead of index, because if any plugin invokes the correct order of reflection before you, then your reflection will fail.

**Comment by ghost** [comment]
P.S.: See this https://stackoverflow.com/questions/25147228/sort-enums-in-declaration-order/25147675#25147675

**Comment by linuxgurugamer** [comment]
The comments made to me were done on the bugtracker, a few years ago and has been cleaned out.  But you can email @jplrepo to ask about it.

From what I can tell, KSP reads all the fields before the mods get to it, so while those articles are valid, I don't think I need to worry.  At least, I haven't had to worry about it for the past 3 years

Thanks

**Comment by ghost** [comment]
Thank you for the knowledge you share with me. I have started my own plugin: https://github.com/firda-cze/TrackingReordered

It is based on this one, but no buttons. It keeps vessels with maneuver on top of the list regardless of filters (had to go deeper in reflection to achieve this), sorts the rest by type then newest first + stops warp as MQ does. That is all I need.

I don't know where will this lead me, but I definitely need more experience on this topic and creating that mod looks like best way for me now. I hope you don't mind that I used ideas from MQ.

Reflect.cs contains all the reflection (versioning can be done in static constructor), will try searching by name first (should work with 64bit windows version for long time), then by index (in MetadataToken-sorted list) and rememberes that to reflect the rest the same way. It will optionaly try the search by type (n-th field of specified type), but will always start with search by index if previous class was resolved by this kind of search. It will print the fields if using the third option and it is disabled by default (indexByType = -1). Will see how good it is when they release new version.

Thanks again.

**Comment by JPLRepo** [comment]
I haven't looked at this code closely. But just wanted you to know:
The use of reflection against the KSP dll's to reference data that is not public is against the EULA.

**Comment by ghost** [comment]
@JPLRepo 7. YOU SHALL NOT:
7.7 Reverse engineer, derive source code, modify, decompile, disassemble, or create derivative works of the Licensed Application, in whole or in part.

Nice, so I have broken EULA just by using ILSpy to see how it all works. Looks like Squad hates modders then.

**Comment by JPLRepo** [comment]
Yes.... there is a public API that is published. There is a forum with which you can raise questions with other modders and the devs.
Modding is encouraged. But you are not supposed to break the rule. Particularly using reflection on non public aspects of the code.

**Comment by linuxgurugamer** [comment]
The way I do this, both with EEX and with the changes in this mod, is to dump the offsets and then hard-code the offsets.
this is what @JPLRepo suggested I do several years ago when I first took over EEX.  I don't know what others did before.
@JPLRepo do you remember the exchange we had back then?  I was new to this, and you almost begged me to not use the names but to use offsets.

**Comment by ghost** [comment]
@JPLRepo They are blocking my registration, don't know why (I was never banned from anywhere, I have bought the game on Steam under account with same user name as I was trying to register there), support is not responding. Seeing real code is much better than public API, that most often tells nothing more than a dump of all public methods and properties.

EULA does not speak about reflecting private properties, it talks about THE SOFTWARE and prohobits even modification. Isn't modding modifying the software? It is not modifying the binares, but it is modifying the software, but I am not a lawyer. I found some question about that from 2014, no response from the devs.

This reminds me of an article stating that Blizzard blocked all SC2 players, that used some kind of teaching tweaks (slowing time) for single-player, because of badges, they say. My mod does no harm (just reorders widgets which are public, but keeping wessels with maneuver on top even if they should be filtered-out seems imposible without reflection of private fields). And it all started with a mod that was already there, so, I really did not expect that. I feel like deleting both repos right now....

**Comment by JPLRepo** [comment]
I am not sure what registration you are talking about. 
Sorry I can not comment further here with regard to what you are asking.
I am a dev, I am not a lawyer, nor do I represent the owner of the copyright material. 
I can only advise you as one dev to another dev/modder that it would be wise to not use reflection on non-public code and that you don't talk about things along those lines. Especially when the forum and the EULA states that you cannot.

**Comment by ghost** [comment]
@JPLRepo Thanks anyway. I was finally able to registr on KSP forums (from a different PC). I will try to search for answers there and keep your words in mind... Maybe I don't really need to access private fields and methods to achieve what I want... maybe, I can just add new window with my own widgets or something like that, I want it in map anyway (at least top few with closest maneuver). I will also check other mods especially Kerbal Alarm Clock to see what that can do. Thank you

**Pull request merged**